### PR TITLE
[com_fields] Add a check if the default value is correct based on JFormRules

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for Issue #15005.

### Summary of Changes
This pr adds support to check the default value of a custom field when a JFormRule exists. Additionally it adds support of rules for field plugins to a be found by JForm.

### Testing Instructions
- Create a calendar custom field
- Set as default value `test`
- Save the field

Please also save the field with the following default values which should be correct:
- now
- NOW
- 2017-02-09

The same can also be tested with any other field which has a rule.

### Expected result
The field is not saved and an error message is shown.

### Actual result
The field saves normally.

### CODE REVIEW

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 2158030fbcadb67838ea60d159d87c22103477d0<br /><br />Test on calendar custom field: "test" brings Error, other 3 Values saved and working in Article.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15032">issues.joomla.org/tracker/joomla-cms/15032</a>.</sub>

File: administrator/components/com_fields/models/field.php [hunk 1]
```diff
@@ -180,6 +187,84 @@ public function save($data)
 		return true;
 	}
 
+
+	/**
+	 * Checks if the default value is valid of the given data. If a string is returned then
```
Comment (brianteeman) [inline administrator/components/com_fields/models/field.php]: For the given data not of

File: libraries/joomla/form/rule/calendar.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
+<?php
+/**
+ * @package     Joomla.Platform
+ * @subpackage  Form
+ *
+ * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE
```
Comment (brianteeman) [inline libraries/joomla/form/rule/calendar.php]: Deploy version

**Comment by Bakual** [comment]
I have tested this item :white_check_mark: successfully on 2158030fbcadb67838ea60d159d87c22103477d0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15032">issues.joomla.org/tracker/joomla-cms/15032</a>.</sub>

File: libraries/joomla/form/rule/calendar.php (hunk 2 above)
Comment (zero-24) [inline libraries/joomla/form/rule/calendar.php]: where do you want a `__DEPLOY_VERSION__` ? I guess you mean line 34?

**Comment by zero-24** [comment]
@laoneo i have just fixed drone.

File: administrator/components/com_fields/models/field.php [hunk 3]
```diff
@@ -195,7 +196,7 @@ public function save($data)
 	 *
 	 * @param   array  $data  The data.
 	 *
-	 * @return  true|string  true if valid, a string when not.
+	 * @return  true|string  true if valid, a string containing the expection message when not.
```
Comment (Quy) [inline administrator/components/com_fields/models/field.php]: Spelling `exception`.

Comment (zero-24) [inline administrator/components/com_fields/models/field.php]: sorry my fault fixed it with https://github.com/joomla/joomla-cms/pull/15032/commits/239402416c592af6989950ee421a6542388cbc4b thanks for spotting.

**Comment by laoneo** [comment]
Thanks!

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 239402416c592af6989950ee421a6542388cbc4b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15032">issues.joomla.org/tracker/joomla-cms/15032</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
@Bakual can you please retest?

**Comment by laoneo** [comment]
Only cs commits are done, guess ready for rtc?

**Comment by franz-wohlkoenig** [comment]
RTC after two successful testes.

**Comment by brianteeman** [comment]
How can this be rtc when there are unresolved comments

**Comment by zero-24** [comment]
Should be fixed now @brianteeman thanks for the reminder.

**Comment by franz-wohlkoenig** [comment]
RTC okay?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15032">issues.joomla.org/tracker/joomla-cms/15032</a>.</sub>

**Comment by laoneo** [comment]
ok now @brianteeman?

**Comment by zero-24** [comment]
Back to RTC after the comment by @brianteeman is resovled now and there are just cs changes between the tests :smile: Thanks :+1:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15032">issues.joomla.org/tracker/joomla-cms/15032</a>.</sub>

**Pull request merged**